creator-node/src/services/stateMachineManager/CNodeHealthManager.ts

```diff
@@ -0,0 +1,318 @@
+import type { StateMonitoringUser } from './stateMonitoring/types'
+
+import axios from 'axios'
+
+import config from '../../config'
+import { logger } from '../../logging'
+
+const redis = require('../../redis')
+const { hasEnoughStorageSpace } = require('../../fileManager')
+
+const PEER_HEALTH_CHECK_REQUEST_TIMEOUT_MS = config.get(
+  'peerHealthCheckRequestTimeout'
+)
+const MINIMUM_MEMORY_AVAILABLE = config.get('minimumMemoryAvailable')
+const MAX_FILE_DESCRIPTORS_ALLOCATED_PERCENTAGE =
+  config.get('maxFileDescriptorsAllocatedPercentage') / 100
+const MINIMUM_DAILY_SYNC_COUNT = config.get('minimumDailySyncCount')
+const MINIMUM_ROLLING_SYNC_COUNT = config.get('minimumRollingSyncCount')
+const MINIMUM_SUCCESSFUL_SYNC_COUNT_PERCENTAGE =
+  config.get('minimumSuccessfulSyncCountPercentage') / 100
+const MAX_STORAGE_USED_PERCENT = config.get('maxStorageUsedPercent')
+
+// Max number of seconds a primary may be unhealthy for since the first time it was seen as unhealthy
+const MAX_NUMBER_SECONDS_PRIMARY_REMAINS_UNHEALTHY = config.get(
+  'maxNumberSecondsPrimaryRemainsUnhealthy'
+)
+// Max number of seconds a secondary may be unhealthy for since the first time it was seen as unhealthy
+const MAX_NUMBER_SECONDS_SECONDARY_REMAINS_UNHEALTHY = config.get(
+  'maxNumberSecondsSecondaryRemainsUnhealthy'
+)
+const REDIS_KEY_PREFIX_PRIMARY = 'earliestFailedHealthCheckPrimary:'
+const REDIS_KEY_PREFIX_SECONDARY = 'earliestFailedHealthCheckSecondary:'
+
+const CNodeHealthManager = {
+  getUnhealthyPeers,
+  isNodeHealthy,
+  determinePeerHealth,
+  isNodeHealthyOrInGracePeriod,
+  getEarliestFailedHealthCheckTimestamp,
+  setHealthCheckTimestampToNow,
+  resetEarliestUnhealthyTimestamp,
+  _queryVerboseHealthCheck,
+  _computeContentNodePeerSet
+}
+
+/**
+ * Performs a health check on the peer set
+ * @param {Object[]} nodeUsers array of objects of schema { primary, secondary1, secondary2, user_id, wallet }
+ * @param {string} contentNodeEndpoint the IP address / URL of this Content Node
+ * @param {boolean?} [performSimpleCheck=false] flag to dictate whether or not to check health check response to
+ *  determine node health
+ * @returns the unhealthy peers in a Set
+ *
+ * @note consider returning healthy set?
+ * TODO - add retry logic to node requests
+ */
+async function getUnhealthyPeers(
+  nodeUsers: StateMonitoringUser[],
+  thisContentNodeEndpoint: string,
+  performSimpleCheck = false
+) {
+  // Compute content node peerset from nodeUsers (all nodes that are in a shared replica set with this node)
+  const peerSet = CNodeHealthManager._computeContentNodePeerSet(
+    nodeUsers,
+    thisContentNodeEndpoint
+  )
+
+  /**
+   * Determine health for every peer & build list of unhealthy peers
+   * TODO: change from sequential to chunked parallel
+   */
+  const unhealthyPeers = new Set()
+
+  for await (const peer of peerSet) {
+    const isHealthy = await CNodeHealthManager.isNodeHealthy(
```

Comment:
should this consume isNodeHealthyOrInGracePeriod because getUnhealthyPeers is also used in monitorState? i guess it would just add the jobs and then not issue the update later

Reply:
I was trying to deviate as little as possible especially since this function already has a couple TODOs like "add retry logic" and "consider returning healthy set." I figure the whole thing will need more rethinking